### DISCUSSION TITLE
Wrapping hyper::body::Incoming using a custom body with custom data

### DISCUSSION THREAD

**Discussion by OldMansKid** [hyperium/hyper#3581] [Q&A] [answered]
Hi everyone,
When implementing hyper::body::Body for use in the context of proxy, I find it very difficult to wrap hyper::body::Incoming if the custom body type doesn't use Bytes. For example:
```
pub enum BodyWrapper { Incoming(hyper::body::Incoming), LocalCache(DataChannel) }
impl hyper::body::Body for BodyWrapper {
    type Data = DataAdapter;
    fn poll_frame(self: Pin<&mut Self>, cx: &mut Context<'_>,)
     -> Poll<Option<Result<Frame<DataAdapter>, Self::Error>>> {
        match self.get_mut() {
            BodyWrapper::Incoming(incoming) => {
                match incoming.frame().poll_unpin(cx)
                ...
            }
            ...
        }
    }
}
pub enum DataAdapter {Owned(Bytes), Shared(ReusableBuf)}
impl Buf for DataAdapter {...}
```
 Because polling `Frame<Incoming>` would get a `Frame<Bytes>`, which doesn't implement `Buf`, nor can it be converted to Bytes either if the data type is trailer. So the only way I can think of to make this compatible when `Frame<Bytes>` is trailers, is to convert the `HeaderMap<HeaderValue>` items one by one into bytes. Although trailers aren't common, there probably won't be much overhead, but is this the only possible way?

**Comment by dswij** [comment, marked as answer]
Are you expecting only to handle data frames? Perhaps you can get the data and then map it: https://docs.rs/hyper/latest/hyper/body/struct.Frame.html#method.into_data

**Reply by OldMansKid** [reply]
Thanks for replying. No, it's a proxy, so there might be trailers. But from my log I think trailers are very rare, and I have been using the proxy I wrote for quite a while. From your reply I suppose there is no other way. I'm going to write code to convert HeaderMap then.

**Reply by seanmonstar** [reply]
It sounds like you want https://docs.rs/hyper/latest/hyper/body/struct.Frame.html#method.map_data

**Reply by OldMansKid** [reply]
I just used into_data() and into_trailers(). To be honest, I've been a little confused by the hyper::body::Body trait, as the poll_frame() seems to only return body data, while hyper::body::Incoming can return both body frames and trailer frames. So I just converted HeaderMap into u8 slice in my Buf implementation, then send trailer along with body anyway.

**Reply by dswij** [reply]
poll_frame returns both data and trailer frames. https://docs.rs/hyper/latest/src/hyper/body/incoming.rs.html#270

But I get that the issue is that you're trying to wrap both data and trailers but the type needs to be converted first.

I wonder if it makes sense to provide some function for `HeaderMap` to convert to Bytes/implement Buf for convenience.

**Reply by OldMansKid** [reply]
Yeah, exactly that. I have to do pretty much the opposite of the parsing done by Frame<Bytes>, like inserting b"0\r\n", just to make the types compatible.

**Reply by sfackler** [reply]
Why are you trying to convert the trailers to bytes? Just make a trailers frame: https://docs.rs/http-body/latest/http_body/struct.Frame.html#method.trailers, or use map_data like Sean said.

**Reply by OldMansKid** [reply]
Sorry, I'm not that experienced in Rust and didn't get Sean's idea earlier. And similarly, I thought `Frame::traiers(t)` would return a `Frame<HeaderMap>` thus incompatible with `Frame<DataAdapter>`, but now I saw it is a generic function and its type isn't determined by the HeaderMap passed in. All good now, I'm going to just use map_data(). Thanks for pointing that out.
